include/exadg/acoustic_conservation_equations/spatial_discretization/spatial_operator.cpp

```diff
@@ -0,0 +1,383 @@
+{
+  return constraint_u;
+}
+
+template<int dim, typename Number>
+dealii::types::global_dof_index
+SpatialOperator<dim, Number>::get_number_of_dofs() const
+{
+  return dof_handler_u.n_dofs() + dof_handler_p.n_dofs();
+}
+
+/*
+ * Initialization of vectors.
+ */
+template<int dim, typename Number>
+void
+SpatialOperator<dim, Number>::initialize_dof_vector(BlockVectorType & dst) const
+{
+  // pressure (1st block) + velocity (2nd block)
+  dst.reinit(2);
+
+  matrix_free->initialize_dof_vector(dst.block(0), get_dof_index_pressure());
+  matrix_free->initialize_dof_vector(dst.block(1), get_dof_index_velocity());
+
+  dst.collect_sizes();
+}
+
+template<int dim, typename Number>
+void
+SpatialOperator<dim, Number>::prescribe_initial_conditions(BlockVectorType & dst,
+                                                           double const      time) const
+{
+  field_functions->initial_solution_pressure->set_time(time);
+  field_functions->initial_solution_velocity->set_time(time);
+
+  // This is necessary if Number == float
+  using VectorTypeDouble = dealii::LinearAlgebra::distributed::Vector<double>;
+
+  VectorTypeDouble pressure_double;
+  VectorTypeDouble velocity_double;
+  pressure_double = dst.block(0);
+  velocity_double = dst.block(1);
+
+  dealii::VectorTools::interpolate(*get_mapping(),
+                                   dof_handler_p,
+                                   *(field_functions->initial_solution_pressure),
+                                   pressure_double);
+
+  dealii::VectorTools::interpolate(*get_mapping(),
+                                   dof_handler_u,
+                                   *(field_functions->initial_solution_velocity),
+                                   velocity_double);
+
+  dst.block(0) = pressure_double;
+  dst.block(1) = velocity_double;
+}
+
+template<int dim, typename Number>
+void
+SpatialOperator<dim, Number>::evaluate(BlockVectorType &       dst,
+                                       BlockVectorType const & src,
+                                       double const            time) const
+{
+  evaluate_acoustic_operator(dst, src, time);
+
+  // shift to the right-hand side of the equation
+  dst *= -1.0;
+
+  apply_inverse_mass_operator(dst, dst);
+}
+
+template<int dim, typename Number>
+void
+SpatialOperator<dim, Number>::evaluate_acoustic_operator(BlockVectorType &       dst,
+                                                         BlockVectorType const & src,
+                                                         double const            time) const
+{
+  acoustic_operator.evaluate(dst, src, time);
+}
+
+template<int dim, typename Number>
+void
+SpatialOperator<dim, Number>::apply_inverse_mass_operator(BlockVectorType &       dst,
+                                                          BlockVectorType const & src) const
+{
+  inverse_mass_pressure.apply(dst.block(0), src.block(0));
+  inverse_mass_velocity.apply(dst.block(1), src.block(1));
```

Comment:
If I remember correctly, we have an `inverse_mass_all` operator in the compressible Navier-Stokes module. The main advantage is probably that in case of deformed elements, geometric mapping data is loaded only once during operator evaluation, and not multiple times for every physical quantity of the block vector.

@kronbichler what do you think? should be introduce such an `inverse_mass_all` also here?

Reply:
I am open for other ways. I just want to mention, that in the compressible INSE module the quantities are not stored in block vectors but in a single vector. The reason why this is not done here, is because we have to be able to access pressure and velocities in aero-acoustics (this can not be seen from this PR). Additionally, it is possible to use mixed orders for pressure and velocity for the acoustic module, which can only be achieved with block vectors (this can also not be seen immediately from this PR).

Reply:
I see, so we would have to figure out whether a combined inverse mass matrix is possible with block-vectors? The polynomial degrees might be different, but in case they are the same it could still be a reasonable optimization to have a combined inverse mass matrix. Or do you think the case of different polynomial degrees will be the standard case?

Reply:
Using mixed degrees might yield a large potential for performance optimization while keeping the same convergence rates, but I didn't investigate this in detail.

If it is possible to have an `inverse_mass_all` operator for block vectors without touching `deal.II`, I think we should make use of it. I would suggest filing an issue for this and proceed with the current implementation in the meantime.